### PR TITLE
DAOS-10894 pool: reject pool connection from old clients

### PR DESCRIPTION
**Pull request by wangshilong**
If pool upgrade succeed for EC rotation, old client(2.2) should
be not allowed to access upgraded pools any more unless clients
are upgraded to 2.4 as well.

Modify pool connection rpc to take max pool version that client
could be supported to server side, server will reject if pool's
version is larger than client max supported version.

Signed-off-by: Wang Shilong <shilong.wang@intel.com>
Required-githooks: true

### CODE REVIEW

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:2359]: If in1->pci_pool_version > DAOS_PROP_PO_GLOBAL_VERSION, that means it is newer clients vs old server, which should be allowed to connect, (in theory, newer clients should be able to access old and new clients.)
the question might be how much delta we allowed here? Eg if pci->pci_pool_version > DAOS_PROP_PO_GLOBAL_VERSION + 1, we also reject.

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:2368]: I think yes, with EC rotation and corresponding upgrade changes, DAOS_POOL_GLOBAL_VERSION will be bumped to 2.

Comment (kccain) [reply src/pool/srv_pool.c:2368]: So at this point this branch will not be taken in the code since it will need to be tested with another patch that will increment DAOS_POOL_GLOBAL_VERSION?

**Review by kccain — CHANGES_REQUESTED** [review]

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:2368]: Yup

**Review by kccain — APPROVED** [review]
I think the C changes are good based on my previous inspection. Approving now, realizing that this PR needs to merge latest master. A lot of the diffs here are already incorporated in this commit just landed to master. So maybe the patch will get a lot smaller after merging.
landed https://github.com/daos-stack/daos/commit/cf2a54f773305cb887446be8926698f65d5641e7

**Review by kccain — DISMISSED** [review]
Looks good, although the different versioning fields and variables are sometimes difficult to read and interpret.

- RPC pci_pool_version is a layout format version (1 currently, 2 later with another patch)
- DAOS_POOL_GLOBAL_VERSION is a layout format version (1, 2 later with another patch)
- DAOS_POOL_VERSION is a pool RPC protocol version (4 or 5)
- and in ds_pool_connect_handler() the handler_version is meant to refer to the RPC version (4 or 5 currently) but takes on the values 0 or 1.

**Comment by wangshilong** [comment]
> Looks good, although the different versioning fields and variables are sometimes difficult to read and interpret.
> 
> * and in ds_pool_connect_handler() the handler_version is meant to refer to the RPC version (4 or 5 currently) but takes on the values 0 or 1.

This is confusing and better fix.

**Review by kccain — DISMISSED** [review]
LGTM

File: src/pool/rpc.h
Comment (kjacque) [inline src/pool/rpc.h:42]: I do find myself wondering how we'll handle this if there are more versions added in the future. These macros could get pretty ugly.

File: src/pool/srv_pool.c
Comment (kjacque) [inline src/pool/srv_pool.c:2360]: We should be checking `in1->pci_pool_version` against the length of the array before indexing into it.

File: src/pool/srv_pool.c
Comment (kjacque) [inline src/pool/srv_pool.c:2362]: I think it would be valuable to put a bit more detail in the error. I.e. what the actual global version is, and what the limitation is.

File: src/pool/srv_pool.c
Comment (kjacque) [inline src/pool/srv_pool.c:2370]: Again would like a more detailed error here

**Review by kjacque — CHANGES_REQUESTED** [review]

File: src/pool/rpc.h
Comment (wangshilong) [reply src/pool/rpc.h:42]: Yes i agreed, but a second think is i think we are going to maintain two versions, that might be fine.

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:2360]: oh, see in1->pci_pool_version <= DAOS_POOL_GLOBAL_VERSION, this is length check? maybe i should add a assertion that pool_max_supported_version length is DAOS_POOL_GLOBAL_VERSION

Comment (wangshilong) [reply src/pool/srv_pool.c:2360]: maybe it is ok, as pool_max_supported_version is defined using macro.
static unsigned int pool_max_supported_version[DAOS_POOL_GLOBAL_VERSION + 1] = {
        1,
        1,
};

**Review by kccain — DISMISSED** [review]

File: src/pool/srv_pool.c
Comment (kjacque) [reply src/pool/srv_pool.c:2360]: Ah, thanks, I missed that. I wonder if it would be valuable to have a separate NUM_POOL_VERSIONS macro. Will it always be global version + 1? Or will that change once we begin bumping the version? Will the array actually grow?

**Review by kjacque — DISMISSED** [review]

**Review by kccain — DISMISSED** [review]

**Review by kjacque — DISMISSED** [review]

**Comment by kprantis** [comment]
Looks like this has a conflict that needs resolving?

**Comment by wangshilong** [comment]
Trying to rebase and also address a missing comments from Kris before.

**Review by kccain — APPROVED** [review]

**Review by kccain — APPROVED** [review]

File: src/pool/srv_pool.c
Comment (kjacque) [inline src/pool/srv_pool.c:2334]: Instead of a numeric literal, shouldn't this be `DAOS_PROP_PO_GLOBAL_VERSION - 1`?

Makes me wonder if the array is the right way to represent this. Essentially we want to make calculations based on the global version to determine supported versions, assuming they are sequential as shown here.

**Review by kjacque — DISMISSED** [review]
Wouldn't block on my comment, just some food for thought.

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:2334]: No, it could be sometimes DAOS_PROP_PO_GLOBAL_VERSION or DAOS_PROP_PO_GLOBAL_VERSION - 1, depends on if there are some in-compatible change which could make client unable to read data from newer pool.

For example, in 2.0/2.2, it is compatible, so they are both 1. but for 2.2/2.4, it will be [2, 1], and maybe in the 2.4/2.6 it might be [3, 2]

Comment (kccain) [reply src/pool/srv_pool.c:2334]: Ah I didn't notice this issue again in this most recent patch. Shouldn't the code be using DAOS_POOL_GLOBAL_VERSION  and DAOS_POOL_GLOBAL_VERSION  -1  (2 and 1)

(and not DAOS_PROP_PO_GLOBAL_VERSION that is a very different value: 14)
?

**Review by kccain — CHANGES_REQUESTED** [review]
I think code needs to change DAOS_PROP_PO_GLOBAL_VERSION  to DAOS_POOL_GLOBAL_VERSION in the definition of max_supported_pool_version[] array

**Comment by wangshilong** [comment]
> I think code needs to change DAOS_PROP_PO_GLOBAL_VERSION to DAOS_POOL_GLOBAL_VERSION in the definition of max_supported_pool_version[] array

Thanks for catching this, fixed, and i've tried to verified by bumping version in the server side:
[root@server daos]# ./install/bin/daos container create --pool pool1 --label cont1
pool ERR  src/pool/cli.c:590 pool_connect_cp() failed to connect to pool: DER_NOTSUPPORTED(-2037): 'Operation not supported'
ERROR: daos: DER_NOTSUPPORTED(-2037): Operation not supported

In the server log:
07/21-21:12:06.38 daos1 DAOS[17081/0/135] pool ERR  src/pool/srv_pool.c:2457 ds_pool_connect_handler() 7f154a78: cannot connect, pool layout version(2) > max client supported pool layout version(1) try to upgrade client firstly.

**Review by kccain — DISMISSED** [review]

File: src/pool/srv_pool.c
Comment (kjacque) [reply src/pool/srv_pool.c:2334]: Good catch on Ken's part. I am still trying to understand this though. Do we want to be chained to updating this array for every release, depending on whether the pool version has been bumped between the latest and previous? It just seems highly likely to be missed or forgotten.

To me it seems practical to do something a bit more generic that isn't coupled so tightly to release versions, e.g. support the current and previous pool version regardless of what the release version says. If we truly want to couple to the release version we should probably be more explicit about it.

Comment (wangshilong) [reply src/pool/srv_pool.c:2334]: I agree this is likely to be missed, i've added comments in the DAOS_POOL_GLOBAL_VERSION to update supported array when you bumped it.

And unfortunately, i did not get way to fix this easily, for example when upgrading DAOS pool from 2.2 to 2.4 with EC rotation enabled, we need upgrade client version as well to access latest pool format, in this case, latest pool version won't work for old clients, then old client should be upgraded to latest version to access new format.

We could not easily say we always support current and previous pool version regardless of what the release version says.

But I guess a safe way is to always reject old clients connection to new pools, that means 2.2 client connect to 2.4 pool won't work, 2.4 client connects to 2.6 pool version won't work.(unless you upgrade client to latest as well). This is reliable but not that friendly, eg 2.0 client actually access 2.2 pool without problem.

And i guess in the future, most of time,  older client could connect new pool without problem, upgrade client as well means we change some fundamental client prototype, i guess this won't happen frequently, so i'd prefer to keep the array now.

What do you think?
Thanks!

Comment (wangshilong) [reply src/pool/srv_pool.c]: A rethink on this, i think maybe it is fine to reject old clients connection with newer pool format is fine.
This is safe and simple for now.

For allowed old client connect to old pool only.

**Comment by wangshilong** [comment]
Tested manually by bumping Version on the server side and created a pool based on that, then try old client and new clients to verify it works as expected:
[root@daos1 daos_test]# ./install/bin/daos container list --pool pool1
pool ERR  src/pool/cli.c:590 pool_connect_cp() failed to connect to pool: DER_NOTSUPPORTED(-2037): 'Operation not supported'
ERROR: daos: failed to connect to pool pool1: DER_NOTSUPPORTED(-2037): Operation not supported
[root@daos1 daos_test]# cd ..
[root@daos1 wc_gerrit]# cd daos
[root@daos1 daos]# ./install/bin/daos container list --pool pool1
No containers.

File: src/pool/srv_pool.c
Comment (kjacque) [reply src/pool/srv_pool.c]: I think that is fine too. Assuming we are meaning it is determined by the pool version not the release version.

**Comment by wangshilong** [comment]
ping reviewers!

File: src/pool/srv_pool.c
Comment (kccain) [reply src/pool/srv_pool.c]: My understanding from a discussion with @johannlombardi was that old client software should be (generally) able to connect to newer server/engine software. The 2.2 client connecting to 2.4 (with pool format updated for EC rotation) I guess is an exception to the general goal. 

So is the plan for now (with this patch) to simply reject older client software attempting to connect to newer pools? And decide to address the general goal later in some future patch?

Comment (wangshilong) [reply src/pool/srv_pool.c]: I think after applying PR, old clients still could connect newer servers if that is old pool, except is if pool is upgraded or newly created, old clients won't connect, @johannlombardi  are you ok with this? or do you think we'd better maintain supported matric which will try best to support old clients connecting new pool if it is possible.

**Comment by kprantis** [comment]
Please prioritize reviews on this 2.2 blocker.

**Review by kccain — APPROVED** [review]
approving PR in its latest state since it has the infrastructure in place to do some more advanced pool connect compatibility logic in the (near?) future if it is decided that the product should have some pool layout compatibility "window" of connection support.

**Review by wangdi1 — APPROVED** [review]

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c]: This is done in the latest codes

**Pull request merged**
